app/models/annotation_merger.rb

```diff
@@ -0,0 +1,121 @@
+class AnnotationMerger
+  def initialize(annotations)
+    # If annotation["text"] is nil, set it to an empty string.
+    # If the text ends with a period, add a space to clarify sentence boundaries when merging annotations.
+    # This prevents loss of readability or meaning.
+    # Setting nil to an empty string also avoids errors in subsequent processing.
+    # Ensure consistent denotations and relations format for reliable merging.
+    @annotations = annotations.map do |annotation|
+      text = annotation["text"] || ""
+      text += " " if text.end_with?(".")
+      denotations = annotation["denotations"] || []
+      relations = annotation["relations"] || []
+      annotation.merge("text" => text, "denotations" => denotations, "relations" => relations)
+    end
+
+    # Pre-build Denotation ID sets
+    denotation_id_sets = @annotations.map { |a| a["denotations"].map { |d| d["id"] }.to_set }
+
+    # Check referential integrity of relations
+    @annotations.each_with_index do |annotation, idx|
+      relations = annotation["relations"]
+      denotation_ids = denotation_id_sets[idx]
+      relations.each do |relation|
+        unless referable_to?(relation, denotation_ids)
```

Comment:
引数の非対称性が設計的に不自然です。
relation は denotation を参照するデータなのに対して、denotation_ids は「IDの集合」に変換された情報であり、引数同士の抽象度がズレている点が気になります。

元データである denotations をそのまま渡すようにすれば、引数の意味が分かりやすくなります。

```suggestion
        unless referable_to?(relation, denotation)
```

このようにしておくと、呼び出し側から見て

- relation は参照しようとしているデータ
- denotations は参照されるデータ

という関係性が明確になり、コードの読みやすさと保守性が高まります。

※ もちろん、referable_to? 側で denotations.map { _1["id"] } のように ID リストに変換する処理は必要になりますが、責務の分離としてはこちらの方が自然です。